### DISCUSSION TITLE
Quarto speed significantly slower than other python IDEs when working with huge dataframes

### DISCUSSION THREAD

**Discussion by hrushikul** [quarto-dev/quarto-cli#2496] [Q&A]
The issue that we see is that Quarto takes a significantly longer time in executing a Python code chunk when compared to running Python script on Rstudio, Jupyter notebook or PyCharm. 
Below is the time taken to execute them on two different windows machines with 8GB RAM and a processing speed of 2.1GHz.

<img width="317" alt="image" src="https://user-images.githubusercontent.com/68015807/191007339-cb6bf111-8c5d-49ee-bc77-5e1e5b559c16.png">

Why is Quarto taking a much longer time? Can you please help us with this? Our goal is to use one system for developing codes using both R and Python

Please Run this code in a **quarto document** inside a **python** chunk
[quarto speed check.txt](https://github.com/quarto-dev/quarto-cli/files/9606815/quarto.speed.check.txt)

**Comment by dragonstyle** [comment]
Without seeing the actual code that is triggering the issue, it is very hard to diagnose the reason. Can you share an example we could use to better understand the issue?

**Reply by hrushikul** [reply]
[quarto speed check.txt](https://github.com/quarto-dev/quarto-cli/files/9603897/quarto.speed.check.txt)

Kindly check the above code attached.
Please Run this code in a **quarto document** inside a **python** chunk

Regards,
Hrushikesh

**Reply by cscheid** [reply]
Thanks for the test case. I can't reproduce the issue. Inside quarto, I get:

<img width="349" alt="image" src="https://user-images.githubusercontent.com/285675/191292996-dcb5f0f6-3894-4bb6-a237-242ae0b1e30f.png">

In the command line, I get:

```
$ python python-perf-test.py
/Users/cscheid/virtualenvs/homebrew-python3/lib/python3.10/site-packages/pandas/util/__init__.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing
/Users/cscheid/Desktop/daily-log/2022/09/20/python-perf-test.py:39: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_old['M'] = df_old['P'] * df_old['Q']
/Users/cscheid/Desktop/daily-log/2022/09/20/python-perf-test.py:53: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['M*Q'] = df_new['P'] * df_new['Q']
--- 2.9333550930023193 seconds ---
```

I suspect the environment you're running quarto is different from the other ones, and it's using different library versions, including ones with performance issues.

**Reply by hrushikul** [reply]
Thank you for your response @cscheid 

I changed my default python environment and set up a new environment especially for running this code as you had suggested. But it still gave me similar results. My python console in the quarto document shows the output as:
<img width="615" alt="image" src="https://user-images.githubusercontent.com/68015807/194539052-eb481ced-fb52-4775-8a69-29ef089c616c.png">

In another iteration it showed the following output:
<img width="619" alt="speed with quarto" src="https://user-images.githubusercontent.com/68015807/194539153-ddb4d2f1-7ca4-46dc-bc8d-8956aa75e4df.png">

The above are the results. It takes more than **140 seconds** in both cases whereas a normal python application like Jupyter notebook takes only around **20-25 sec** for the execution of the same code. Following is the output on the Jupyter - notebook display:
<img width="643" alt="speed with Jupyter notebook" src="https://user-images.githubusercontent.com/68015807/194540819-60ea803b-3eda-4a4e-9ab1-7089c1b2c35e.png">

 I am using a system with **8 GB RAM** and **2.10 GHz processor** and **windows** working environment. I am wondering if you are getting the result in close to **3 seconds**  may have been using a really high-end system with high RAM and processor capacity or maybe a MAC environment. Can you please confirm? Because loading a CSV file with 600,000 rows and 40 columns alone take more than 3 seconds. I also tried the same code with default settings on 4 other laptops with configurations the same as mine. They are getting very similar results.

I have tried quite a few ways for a long time and nothing seems to be working. Can you please point me in the right direction? Am I missing out on something?

**Reply by cscheid** [reply]
You wrote:

> The above are the results. It takes more than 140 seconds in both cases whereas a normal python application like Jupyter notebook takes only around 20-25 sec for the execution of the same code. Following is the output on the Jupyter - notebook display:

I continue to believe that this is not a _quarto_ issue, especially since the warning you are getting doesn't come from quarto, it comes from pandas. As I said in my previous message,

> I suspect the environment you're running quarto is different from the other ones, and it's using different library versions, including ones with performance issues.

You should start by printing out the versions of the libraries you're using.

**Reply by hrushikul** [reply]
Hi @cscheid 

I did what you said. I **updated the libraries** to the latest updates and also **fixed the warning from pandas**. Still, I see a difference in the speeds of execution of the same script in the quarto document Vs (Jupyter notebook, pycharm, and anaconda spyder). All the python compilers work **significantly faster than the quarto** for the same code. [My hypothesis is **python runs slow in quarto for large data frames**]

This is the updated version of the code I am using:
[quarto speed check.txt](https://github.com/quarto-dev/quarto-cli/files/9762570/quarto.speed.check.txt)

Screenshot of Updated libraries in different compilers:
**Jupyter notebook**
<img width="265" alt="jupyter libraries version" src="https://user-images.githubusercontent.com/68015807/195278217-fc3598ec-f25b-4aea-9523-07d032497669.png">
**Pycharm**:
<img width="265" alt="pycharm libraries version" src="https://user-images.githubusercontent.com/68015807/195278247-702466d6-8047-4d83-a751-3bd97fa86079.png">
Updated library version in **quarto document**:
<img width="288" alt="quarto libraries version" src="https://user-images.githubusercontent.com/68015807/195278341-f40e2060-03b9-4706-8cd2-18f8c324da0e.png">

The following is the output time in seconds for different python compilers:
**Jupyter notebook** - time: **23.5 sec**
<img width="455" alt="jupyter console output" src="https://user-images.githubusercontent.com/68015807/195278581-824311ed-a528-4bec-a152-5cbf0c1563eb.png">
**Pycharm** time: **23.5 sec**
<img width="376" alt="pycharm console output" src="https://user-images.githubusercontent.com/68015807/195278642-3e06ca02-ff44-4f9f-b89a-94a7526bfe9b.png">
**Anaconda Spyder** time: **25.8 sec**
<img width="458" alt="spyder libraries version and output" src="https://user-images.githubusercontent.com/68015807/195278741-2cb2e3de-60b9-488e-be3f-7af2dbdbe95f.png">
**Quarto document** time: **111.6 sec**
<img width="463" alt="quarto console iteration 2" src="https://user-images.githubusercontent.com/68015807/195278806-dff90ca5-581a-45bd-85e1-78f8bc0b009e.png">


There clearly seems to be a difference between the working times of python in python compilers vs python in the quarto document. I strongly feel it may be because **python runs slow in the quarto document when the data frames it is handling is large**.

Can you please confirm the same?

**Reply by cscheid** [reply]
I still can't reproduce it. On my side, the quarto version:

<img width="865" alt="image" src="https://user-images.githubusercontent.com/285675/195355480-f7536b0a-7f48-49b5-91c1-5319ebe508f9.png">

and the version I run directly from python takes

``` 
$ python speed.py
1.4.4
1.23.2
done
--- 2.965824842453003 seconds ---
```

Your theory for what's happening unfortunately does not match with how quarto execution happens. _Quarto runs python through jupyter_. We literally don't run Python code: we take the python code in a .qmd file, prepare a jupyter notebook, and execute that instead. That's why I said before that this must be some configuration issue with the python environment that we are detecting. It cannot be "quarto executing the code differently" because quarto does not execute python code.

The next thing for you to try is to see if the libraries and implementation you're using are _exactly_ the same. Try the following:

```
$ python
python
Python 3.10.6 (main, Aug 30 2022, 04:58:14) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import platform
>>> import sys
>>> import pandas
>>> sys.modules['pandas']
<module 'pandas' from '/Users/cscheid/virtualenvs/homebrew-python3/lib/python3.10/site-packages/pandas/__init__.py'>
>>> sys.modules['numpy']
<module 'numpy' from '/Users/cscheid/virtualenvs/homebrew-python3/lib/python3.10/site-packages/numpy/__init__.py'>
>>> platform.python_version()
'3.10.6'
>>> platform.uname()
uname_result(system='Darwin', node='Lydian.local', release='21.6.0', version='Darwin Kernel Version 21.6.0: Wed Aug 10 14:28:23 PDT 2022; root:xnu-8020.141.5~2/RELEASE_ARM64_T6000', machine='arm64')
```

Check if the results of the module path match, as well as the details of the platform.

**Comment by hrushikul** [comment]
Hi @cscheid,

The path quarto was using was different from what jupyter notebook was using. The following is the **previous path quarto** was using:
  **Previous path Quarto**  
![quarto python previous path](https://user-images.githubusercontent.com/68015807/197174542-464674ae-c2b2-4da9-b7c9-4d00cb1a005b.png)

I used the following code to change the path:
  **Code for changing the python-path**
<img width="370" alt="quarto python path change" src="https://user-images.githubusercontent.com/68015807/197183975-5a40d438-fe0b-4f5b-8d36-20cb595200ec.png">

The updated quarto path was the following:
  **Updated quarto path**
<img width="488" alt="quarto python new path" src="https://user-images.githubusercontent.com/68015807/197180929-edbc35a8-a106-460a-955a-e24b18c7f9df.png">
  
The only thing is it is showing Python~1 [red underline] instead of Python310 as mentioned in the path. Is this the problem? If yes, how could we change it to the correct path?

  The jupyter path also showed was same:
<img width="815" alt="jupyter python path" src="https://user-images.githubusercontent.com/68015807/197180978-4764d82e-02a6-48d3-9666-ac136c1584ce.png">
  
  But after running the code, the time taken by quarto is still the same:
  **Quarto time with updated python path: 122 sec**
<img width="434" alt="quarto run time" src="https://user-images.githubusercontent.com/68015807/197175253-8abb65f2-5f2a-40b1-ad48-a368b1106c4b.png">
  
  It still didn't change much! Quarto is still taking a long time for the execution of the same code. I am unable to figure out the error. Please let me know your thoughts.

Another thing I wanted to point out is that PyCharm used the same python path and version as what quarto used earlier.
**Pycharm python version**
<img width="545" alt="pycharm python version" src="https://user-images.githubusercontent.com/68015807/197183710-06f9d9ac-fd39-4199-9769-9eac08740e62.png">

But still, the execution time of Pycharm for the same code is much faster than quarto [old python path]
**PyCharm execution time: 16 sec**
<img width="568" alt="pycharm execution time" src="https://user-images.githubusercontent.com/68015807/197182005-4b20fd2e-d3a9-4e6c-b7cd-8736f0c79820.png">

If the python path would've been the error, it should've made the PyCharm IDE work similarly to the quarto in terms of speed of the execution. Still, PyCharm works very similar to Jupyter in terms of speed execution despite having different python paths.

So, is this a Quarto issue in particular or something else?

**Reply by cscheid** [reply]
Check that the _library versions_ of pandas in PyCharm, Jupyter, and quarto are the same. You haven't checked where Jupyter and PyCharm are getting Pandas from.

Also, do not use reticulate for changing the python path. Using that means that you will change the entire engine of execution away from jupyter, rendering the results incomparable.

You will need to figure out what it is in your environment that is causing quarto to find Python 3.9 instead of 3.10 (and, in addition, like I said above, make sure that the _libraries_ are the same.)

**Reply by jthomasmock** [reply]
Hi @hrushikul as Carlos mentioned - there are a few things here that could be different across your speed testing.

You are executing Python code natively in Pycharm/Jupyter, but then you are choosing to switch to `reticulate` to execute the Python code in a Quarto document. Quarto can natively execute Python code via Jupyter kernel (which quite literally uses Jupyter to do ALL of the execution, Quarto doesn't have it's own 'engine'). Quarto can also use `reticulate`, but then there is the overhead of running Python code from a wrapper in R.

In your YAML, you can choose which engine to use:

```
engine: knitr # <- this is forcing R + Reticulate to execute
jupyter: python3 # <- this forces native execution in a python3 Jupyter kernel
```

On M1 Mac - I am getting ~3-5 seconds using either R + `reticulate` to execute your Python code or natively executing the Python code in Python with a Jupyter engine. I realize my environment is different than yours but to Carlos' point - this is not likely to be a Quarto issue but rather a difference in your environment regarding Jupyter kernels and/or `reticulate`.

A check you could do is run your example code in R alone with `reticulate` (option 1), in an RMarkdown file (using `reticulate`) (option 2) and also running your code in Quarto with `jupyter: python3` (option 3).

These results would be an indication that `reticulate` might face slowdowns in your environment (regardless of whether it's used in Quarto or not).

Executing the code in a R script or RMarkdown gives us another indication that perhaps there is some issue you are facing with `reticulate` itself (which you could then move to the `reticulate` Github repo).

The expectation is that natively executing Python code in Quarto via a Jupyter kernel (instead of `reticulate`) gives you a similar speed to Jupyter alone since it literally IS Jupyter executing the code behind the scenes.

Examples below, executing around 3-5 seconds across the various methods with:

Python 3.10.6
Pandas 1.5.0
Numpy 1.23.2

## Reticulate alone, from .R file

<details><summary><code>reticulate::source_python()</code></summary>

Use `reticulate` alone, from a .R file

### `reticulate-speed-test.R`

```r
library(reticulate)

source_python("py-speed-test.py")
```

Then a separate `py-speed-test.py` file.

### `py-speed-test.py`

```python
#################################################### CODE #########################################################
import time
import numpy as np
import pandas as pd
import os
import sys
from platform import python_version
from pandas._testing import rands_array
pd.set_option('display.max_columns',None)

python_version()

# creating a numbers dataframe
alphabets = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
x = pd.DataFrame(np.random.randint(100,10000,size=(600000,26)),columns=list(alphabets))
x.head()

# adding strings to numbers dataframe
N = (600000,14)
s_df = pd.DataFrame(rands_array(10, N))
s_df.shape


# combining number and string dataframe
df = pd.concat([x,s_df],axis=1)
df.shape

# write it to .csv file
path = os.getcwd().replace("\\",'/')
df.to_csv(path+'/quarto_file.csv',index=False)


start_time = time.time()

# reading the data
path = os.getcwd().replace('\\','/')
data = pd.read_csv(path+'/quarto_file.csv')

df_prep = data.copy()

# old data [slicing original data]
df_old = df_prep[df_prep.A>900]
df_old['M'] = df_old['P'] * df_old['Q']
old_group = df_old.groupby(['0']).agg({'G':'sum','Q':'sum','M':'sum','S':'sum'}).reset_index().rename(columns={'G':'old_G','Q':'old_Q','S':'old_S'})
old_group['old_M'] = old_group['M']/old_group['old_Q']
old_group = old_group.drop(['M'],axis=1)
old_group.shape
old_group.columns
old_group['old_G'] = old_group['old_G']/12
old_group['old_Q'] = old_group['old_Q']/12
old_group['old_S'] = old_group['old_S']/12
old_group

# new data [slicing original data]
df_new = df_prep[df_prep.Y>5000]
df_new.shape
df_new['M*Q'] = df_new['P'] * df_new['Q']

new_group = df_new.groupby(['0']).agg({'G':'sum','Q':'sum','M*Q':'sum','S':'sum'}).reset_index().rename(columns={'G':'new_G','Q':'new_Q','S':'new_S'})
new_group['new_M'] = new_group['M*Q']/new_group['new_Q']
new_group = new_group.drop(['M*Q'],axis=1)
new_group.shape
new_group['new_G'] = new_group['new_G']/4
new_group['new_Q'] = new_group['new_Q']/4
new_group['new_S'] = new_group['new_S']/4

# data checks [old data vs new data]
check = pd.merge(left=new_group,right=old_group,how='left',on=['0'])
check.columns
check.shape

check['%M change'] = 100*(new_group['new_M'] - old_group['old_M'])/ old_group['old_M']
check['%G change'] = 100*(new_group['new_G'] - old_group['old_G'])/ old_group['old_G']
check['%Q change'] = 100*(new_group['new_Q'] - old_group['old_Q'])/ old_group['old_Q']
check['%S change'] = 100*(new_group['new_S'] - old_group['old_S'])/ old_group['old_S']


print("--- %s seconds ---" % (time.time() - start_time))
```

</details>

## RMarkdown - Python from R

This is RMarkdown to use `reticulate` - executing Python code from within R.

<details><summary>RMarkdown Example</summary>

This is RMarkdown to use `reticulate` - executing Python code from within R.

File name: `rmd-speed.Rmd`

````
---
output: html_document
engine: knitr
---

Quarto speed check

```{python}
#################################################### CODE #########################################################
import time
import numpy as np
import pandas as pd
import os
import sys
from platform import python_version
from pandas._testing import rands_array
pd.set_option('display.max_columns',None)

python_version()
```

```{python}
# creating a numbers dataframe
alphabets = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
x = pd.DataFrame(np.random.randint(100,10000,size=(600000,26)),columns=list(alphabets))
x.head()

# adding strings to numbers dataframe
N = (600000,14)
s_df = pd.DataFrame(rands_array(10, N))
s_df.shape


# combining number and string dataframe
df = pd.concat([x,s_df],axis=1)
df.shape

# write it to .csv file
path = os.getcwd().replace("\\",'/')
df.to_csv(path+'/quarto_file.csv',index=False)


start_time = time.time()

# reading the data
path = os.getcwd().replace('\\','/')
data = pd.read_csv(path+'/quarto_file.csv')

df_prep = data.copy()

# old data [slicing original data]
df_old = df_prep[df_prep.A>900]
df_old['M'] = df_old['P'] * df_old['Q']
old_group = df_old.groupby(['0']).agg({'G':'sum','Q':'sum','M':'sum','S':'sum'}).reset_index().rename(columns={'G':'old_G','Q':'old_Q','S':'old_S'})
old_group['old_M'] = old_group['M']/old_group['old_Q']
old_group = old_group.drop(['M'],axis=1)
old_group.shape
old_group.columns
old_group['old_G'] = old_group['old_G']/12
old_group['old_Q'] = old_group['old_Q']/12
old_group['old_S'] = old_group['old_S']/12
old_group

# new data [slicing original data]
df_new = df_prep[df_prep.Y>5000]
df_new.shape
df_new['M*Q'] = df_new['P'] * df_new['Q']

new_group = df_new.groupby(['0']).agg({'G':'sum','Q':'sum','M*Q':'sum','S':'sum'}).reset_index().rename(columns={'G':'new_G','Q':'new_Q','S':'new_S'})
new_group['new_M'] = new_group['M*Q']/new_group['new_Q']
new_group = new_group.drop(['M*Q'],axis=1)
new_group.shape
new_group['new_G'] = new_group['new_G']/4
new_group['new_Q'] = new_group['new_Q']/4
new_group['new_S'] = new_group['new_S']/4

# data checks [old data vs new data]
check = pd.merge(left=new_group,right=old_group,how='left',on=['0'])
check.columns
check.shape

check['%M change'] = 100*(new_group['new_M'] - old_group['old_M'])/ old_group['old_M']
check['%G change'] = 100*(new_group['new_G'] - old_group['old_G'])/ old_group['old_G']
check['%Q change'] = 100*(new_group['new_Q'] - old_group['old_Q'])/ old_group['old_Q']
check['%S change'] = 100*(new_group['new_S'] - old_group['old_S'])/ old_group['old_S']


print("--- %s seconds ---" % (time.time() - start_time))
```
````
</details>

### Quarto: Python Native

This is Quarto, natively executing the Python code via a Jupyter kernel (no R involved)

<details><summary>Quarto Example w/ Jupyter kernel as engine</summary>

This is Quarto, natively executing the Python code via a Jupyter kernel (no R involved)

File name: `qmd-speed-jupyter-kernel.qmd`

````
---
format: html
jupyter: python3
---

Quarto speed check

```{python}
#################################################### CODE #########################################################
import time
import numpy as np
import pandas as pd
import os
import sys
from platform import python_version
from pandas._testing import rands_array
pd.set_option('display.max_columns',None)

python_version()
```

```{python}
# creating a numbers dataframe
alphabets = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
x = pd.DataFrame(np.random.randint(100,10000,size=(600000,26)),columns=list(alphabets))
x.head()

# adding strings to numbers dataframe
N = (600000,14)
s_df = pd.DataFrame(rands_array(10, N))
s_df.shape


# combining number and string dataframe
df = pd.concat([x,s_df],axis=1)
df.shape

# write it to .csv file
path = os.getcwd().replace("\\",'/')
df.to_csv(path+'/quarto_file.csv',index=False)


start_time = time.time()

# reading the data
path = os.getcwd().replace('\\','/')
data = pd.read_csv(path+'/quarto_file.csv')

df_prep = data.copy()

# old data [slicing original data]
df_old = df_prep[df_prep.A>900]
df_old['M'] = df_old['P'] * df_old['Q']
old_group = df_old.groupby(['0']).agg({'G':'sum','Q':'sum','M':'sum','S':'sum'}).reset_index().rename(columns={'G':'old_G','Q':'old_Q','S':'old_S'})
old_group['old_M'] = old_group['M']/old_group['old_Q']
old_group = old_group.drop(['M'],axis=1)
old_group.shape
old_group.columns
old_group['old_G'] = old_group['old_G']/12
old_group['old_Q'] = old_group['old_Q']/12
old_group['old_S'] = old_group['old_S']/12
old_group

# new data [slicing original data]
df_new = df_prep[df_prep.Y>5000]
df_new.shape
df_new['M*Q'] = df_new['P'] * df_new['Q']

new_group = df_new.groupby(['0']).agg({'G':'sum','Q':'sum','M*Q':'sum','S':'sum'}).reset_index().rename(columns={'G':'new_G','Q':'new_Q','S':'new_S'})
new_group['new_M'] = new_group['M*Q']/new_group['new_Q']
new_group = new_group.drop(['M*Q'],axis=1)
new_group.shape
new_group['new_G'] = new_group['new_G']/4
new_group['new_Q'] = new_group['new_Q']/4
new_group['new_S'] = new_group['new_S']/4

# data checks [old data vs new data]
check = pd.merge(left=new_group,right=old_group,how='left',on=['0'])
check.columns
check.shape

check['%M change'] = 100*(new_group['new_M'] - old_group['old_M'])/ old_group['old_M']
check['%G change'] = 100*(new_group['new_G'] - old_group['old_G'])/ old_group['old_G']
check['%Q change'] = 100*(new_group['new_Q'] - old_group['old_Q'])/ old_group['old_Q']
check['%S change'] = 100*(new_group['new_S'] - old_group['old_S'])/ old_group['old_S']


print("--- %s seconds ---" % (time.time() - start_time))
```
````
</details>
